osf/migrations/0164_add_guardian_to_nodes.py

```diff
@@ -0,0 +1,200 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.13 on 2018-08-27 21:07
+from __future__ import unicode_literals
+
+from math import ceil
+
+import logging
+from django.db import migrations, connection
+
+logger = logging.getLogger(__name__)
+
+"""
+NODE MIGRATION
+1) Creates three django groups for each existing abstract node (admin/write/read)
+2) Gives admin groups admin/write/read perms, write groups write/read, and read: read
+ - Populates osf NodeGroupObjectPermission (DFK) table instead of out-of-the-box guardian GroupObjectPermission table
+3) Adds node contributors to corresponding django groups - a node write contributor is added to the node's write django group
+ - Populates OSFUserGroups table with group id/user id pair
+"""
+
+def reverse_guardian_migration(state, schema):
+    sql = """
+        -- Drop NodeGroupObjectPermission table - table gives node django groups
+        -- permissions to node
+        DELETE FROM osf_nodegroupobjectpermission;
+
+        -- Remove user membership in Node read/write/admin Django groups
+        DELETE FROM osf_osfuser_groups
+        WHERE group_id IN (
+          SELECT id
+          FROM auth_group
+          WHERE name LIKE '%' || 'node_' || '%'
+        );
+
+        -- Remove admin/write/read node django groups
+        DELETE FROM auth_group
+        WHERE name LIKE '%' || 'node_' || '%';
+        """
+
+    with connection.cursor() as cursor:
+        cursor.execute(sql)
+
+# Forward migration - for each node, create a read, write, and admin Django group
+add_node_read_write_admin_auth_groups = """
+    INSERT INTO auth_group (name)
+    (SELECT 'node_' || N.id || '_read' FROM osf_abstractnode AS N WHERE N.id > {start} AND N.id <= {end}
+    UNION
+    SELECT 'node_' || N.id || '_write' FROM osf_abstractnode AS N WHERE N.id > {start} AND N.id <= {end}
+    UNION
+    SELECT 'node_' || N.id || '_admin' FROM osf_abstractnode AS N WHERE N.id > {start} AND N.id <= {end});
+    """
+
+# Forward migration - add read permissions to all node django read groups, add read/write perms
+# to all node django write groups, and add read/write/admin perms to all node django admin groups
+add_permissions_to_node_groups = """
+    -- Adds "read_node" permissions to all Node read groups - uses NodeGroupObjectPermission table
+    INSERT INTO osf_nodegroupobjectpermission (content_object_id, group_id, permission_id)
+    SELECT N.id as content_object_id, G.id as group_id, PERM.id AS permission_id
+    FROM osf_abstractnode AS N, auth_group G, auth_permission AS PERM
+    WHERE G.name = 'node_' || N.id || '_read'
+    AND PERM.codename = 'read_node'
+    AND N.id > {start}
+    AND N.id <= {end};
+
+    -- Adds "read_node" and "write_node" permissions to all Node write groups
```

Comment:
Can these two queries be consolidated with: 
```
AND (
  PERM.codename = 'read_node' OR
  PERM.codename = 'write_node'
) [...]
``` 
?
(Disclaimer: I have not checked to see if this works and is faster)

Reply:
Made this change, seems to work, no major changes noticed in speed.